src/physics/circular_filament.rs

```diff
@@ -402,67 +548,597 @@ pub fn vector_potential_circular_filament_par(
+/// This variant is parallelized over chunks of observation points.
+///
+/// # Arguments
+///
+/// * `rzifil`:    (m, m, A-turns) r-coord, z-coord, and current of filament, each length `m`
+/// * `xyzobs`:    (m) Observation point coords, each length `n`
+/// * `jobs`:      (A/m^2) Current density vector at observation point, each length `n`
+/// * `out`:        (N/m^3) Body force density in cartesian form, each length `n`
+pub fn body_force_density_circular_filament_cartesian_par(
+    rzifil: (&[f64], &[f64], &[f64]),
+    xyzobs: (&[f64], &[f64], &[f64]),
+    jobs: (&[f64], &[f64], &[f64]),
+    out: (&mut [f64], &mut [f64], &mut [f64]),
+) -> Result<(), &'static str> {
+    // Chunk inputs
+    let n = chunksize(xyzobs.0.len());
+    let (xpc, ypc, zpc) = par_chunks_3tup!(xyzobs, n);
+    let (jxc, jyc, jzc) = par_chunks_3tup!(jobs, n);
+    let (outxc, outyc, outzc) = mut_par_chunks_3tup!(out, n);
+
+    // Run calcs
+    outxc
+        .zip(outyc.zip(outzc.zip(xpc.zip(ypc.zip(zpc.zip(jxc.zip(jyc.zip(jzc))))))))
+        .try_for_each(|(outx, (outy, (outz, (xp, (yp, (zp, (jx, (jy, jz))))))))| {
+            body_force_density_circular_filament_cartesian(
+                rzifil,
+                (xp, yp, zp),
+                (jx, jy, jz),
+                (outx, outy, outz),
+            )
+        })?;
+
+    Ok(())
+}
+
 #[cfg(test)]
 mod test {
-    use std::f64::consts::PI;
+    use core::f64::consts::PI;
 
     use super::*;
+    use crate::{physics::linear_filament::body_force_density_linear_filament, testing::*};
+
+    /// Make sure that force between a circular filament and a piecewise linear filament
+    /// is equal and opposite
+    #[test]
+    fn test_body_force_density() {
+        // Because the force from the circular filament to the lienar filament is calculated
+        // with closed-form circular filament B-field while the force from the linear filament
+        // to the circular filament is calculated by discretizing the circular filament,
+        // tightening tolerances requires excessive discretization of the circular loop,
+        // and some deviation is expected.
+        let (rtol, atol) = (5e-2, 1e-9);
+
+        // Make some circular filaments
+        let (rfil, zfil, nfil) = example_circular_filaments();
+        // Use number-of-turns as the filament current
+        // so that the result is in per-amp units
+        let rzifil = (&rfil[..], &zfil[..], &nfil[..]);
+
+        // Make a slightly tilted helical piecewise-linear filament
+        let xyzfil1 = example_helix();
+        let n = xyzfil1.0.len();
+        let xyzobs = (
+            &xyzfil1.0[..n - 1],
+            &xyzfil1.1[..n - 1],
+            &xyzfil1.2[..n - 1],
+        );
+        let (x, y, z) = &xyzfil1;
+        let dl = (&diff(x)[..], &diff(y)[..], &diff(z)[..]);
+
+        // Calculate force from circular filaments to helix,
+        // using filament direction vector as the current density vector
```

Comment:
I find the use of dl as the current density in this test to be confusing. It would be clearer to me if this test:
- Assigned an arbitrary cross section area to the conductors
- Calculated `j` as `i / area * magnitude(dl)`
- Calculated the total force [N] on each conductor as the sum of `force_density * area * magnitude(dl)` over its segments
- Asserted that the total forces were equal and opposite from the two calculation methods

That would be mathematically the same as how the test is presently written. But IMO it would be a lot easier to read and see that the units work out.